pandas/_libs/tslib.pyx

```diff
@@ -1266,6 +1266,18 @@ cdef class _Timestamp(datetime):
 
             return result
 
+    property _short_repr:
+        def __get__(self):
+            # format a Timestamp with only _date_repr if possible
+            # otherwise _repr_base
+            if (self.hour == 0 and
+                self.minute == 0 and
+                self.second == 0 and
+                self.microsecond == 0 and
+                self.nanosecond == 0):
+                return self._date_repr
+            return self._repr_base
```

Comment:
Not sure if there's a good way to factor it out, but this is duplicative with the existing formatting stuff here:
https://github.com/pandas-dev/pandas/blob/ad3d88600825d02f4540e2c2614f0c7a93e4af35/pandas/_libs/tslib.pyx#L1791